drivers/bus/pci/pdo.c

```diff
@@ -359,9 +360,10 @@ PdoGetRangeLength(PPDO_DEVICE_EXTENSION DeviceExtension,
              ? (OriginalValue.Bar & PCI_ADDRESS_IO_ADDRESS_MASK_64)
              : (OriginalValue.Bar & PCI_ADDRESS_MEMORY_ADDRESS_MASK_64));
 
-    *Length = ~((NewValue.Bar & PCI_ADDRESS_IO_SPACE)
-                ? (NewValue.Bar & PCI_ADDRESS_IO_ADDRESS_MASK_64)
-                : (NewValue.Bar & PCI_ADDRESS_MEMORY_ADDRESS_MASK_64)) + 1;
+    Size = (NewValue.Bar & PCI_ADDRESS_IO_SPACE)
+           ? (NewValue.Bar & PCI_ADDRESS_IO_ADDRESS_MASK_64)
+           : (NewValue.Bar & PCI_ADDRESS_MEMORY_ADDRESS_MASK_64);
+    *Length = Size & ~(Size - 1);
```

Comment:
This returns the least significant set bit of Size. What's the logic behind that?

Reply:
The BAR size is determined by the position of the least significant read/write bit. Some devices return zeros for upper bits, the patch solves this problem. Also the high part of the resource length must be zero for 32-bit BARs.